### PR TITLE
Add aliases for common DataSource declarations

### PR DESCRIPTION
**Pull request by alexserov**
generator: https://github.com/alexserov/datasource-ts-aliases/

### CODE REVIEW

File: js/ui/collection/ui.collection_widget.base.d.ts [hunk 1]
```diff
@@ -33,9 +31,10 @@ export interface CollectionWidgetOptions<TComponent> extends WidgetOptions<TComp
     /**
      * @docid
      * @default null
+     * @type Store|DataSource|DataSourceOptions|string|Array<any>
```
Comment (RomanTsukanov) [inline js/ui/collection/ui.collection_widget.base.d.ts]: I suppose it should be Array<string | CollectionWidgetItem>

File: js/ui/context_menu.d.ts [hunk 2]
```diff
@@ -98,7 +93,7 @@ export interface dxContextMenuOptions extends dxMenuBaseOptions<dxContextMenu> {
      * @default null
      * @public
      */
-    dataSource?: string | Array<Item> | Store | DataSource | DataSourceOptions;
+    dataSource?: DataSourceDefinition<string | Item>;
```
Comment (RomanTsukanov) [inline js/ui/context_menu.d.ts]: According to the doccomment, the array cannot contain strings.

File: js/ui/context_menu/ui.menu_base.d.ts [hunk 3]
```diff
@@ -52,8 +47,9 @@ export interface dxMenuBaseOptions<TComponent> extends HierarchicalCollectionWid
      * @docid
      * @default null
      * @public
+     * @type Store|DataSource|DataSourceOptions|string|Array<any>
      */
-    dataSource?: string | Array<dxMenuBaseItem> | Store | DataSource | DataSourceOptions;
+    dataSource?: DataSourceDefinition<string | dxMenuBaseItem>;
```
Comment (RomanTsukanov) [inline js/ui/context_menu/ui.menu_base.d.ts]: According to the previous type, the array cannot contain strings.

File: js/ui/context_menu/ui.menu_base.d.ts [hunk 4]
```diff
@@ -52,8 +47,9 @@ export interface dxMenuBaseOptions<TComponent> extends HierarchicalCollectionWid
      * @docid
      * @default null
      * @public
+     * @type Store|DataSource|DataSourceOptions|string|Array<any>
```
Comment (RomanTsukanov) [inline js/ui/context_menu/ui.menu_base.d.ts]: I think this doccomment should be fixed according to the previous type.

File: js/ui/data_grid.d.ts [hunk 5]
```diff
@@ -2722,8 +2722,10 @@ export interface ColumnLookup {
    * @docid GridBaseColumn.lookup.dataSource
    * @type_function_param1_field1 data:object
    * @default undefined
+   * @type_function_return Array<any>|Store|DataSourceOptions
+   * @type Array<any>|Store|DataSourceOptions
    */
-  dataSource?: Array<any> | DataSourceOptions | Store | ((options: { data?: any; key?: any }) => Array<any> | DataSourceOptions | Store);
+  dataSource?: FilterLookupDataSource<any> | (() => FilterLookupDataSource<any>);
```
Comment (dxBeardedBear) [inline js/ui/data_grid.d.ts]: Add argument object for function: options: { data?: any; key?: any }

File: js/ui/data_grid.d.ts [hunk 6]
```diff
@@ -2685,9 +2684,10 @@ export interface ColumnHeaderFilter {
    * @docid GridBaseColumn.headerFilter.dataSource
    * @type_function_param1_field1 component:object
    * @default undefined
-   * @type_function_return void
+   * @type_function_return Array<any>|Store|DataSourceOptions
+   * @type Array<any>|Store|DataSourceOptions
    */
-  dataSource?: Array<any> | Store | ((options: { component?: any; dataSource?: DataSourceOptions }) => any) | DataSourceOptions;
+  dataSource?: FilterLookupDataSource<any> | (() => FilterLookupDataSource<any>);
```
Comment (dxBeardedBear) [inline js/ui/data_grid.d.ts]: Add argument object for function: options: { data?: any; key?: any }

File: js/ui/data_grid.d.ts [hunk 7]
```diff
@@ -2685,9 +2684,10 @@ export interface ColumnHeaderFilter {
    * @docid GridBaseColumn.headerFilter.dataSource
    * @type_function_param1_field1 component:object
    * @default undefined
-   * @type_function_return void
+   * @type_function_return Array<any>|Store|DataSourceOptions
+   * @type Array<any>|Store|DataSourceOptions
```
Comment (dxBeardedBear) [inline js/ui/data_grid.d.ts]: Add Function to @type

File: js/ui/data_grid.d.ts [hunk 8]
```diff
@@ -2722,8 +2722,10 @@ export interface ColumnLookup {
    * @docid GridBaseColumn.lookup.dataSource
    * @type_function_param1_field1 data:object
    * @default undefined
+   * @type_function_return Array<any>|Store|DataSourceOptions
+   * @type Array<any>|Store|DataSourceOptions
```
Comment (dxBeardedBear) [inline js/ui/data_grid.d.ts]: Add Function to @type

File: js/ui/data_grid.d.ts (hunk 6 above)
Comment (RomanTsukanov) [inline js/ui/data_grid.d.ts]: AFAIK, there's no need to return anything from this function.

https://js.devexpress.com/Documentation/21_2/Guide/UI_Components/DataGrid/How_To/Customize_Header_Filter_Data_Source/#Change_the_Generated_Data_Source

File: js/viz/vector_map.d.ts [hunk 9]
```diff
@@ -365,7 +363,7 @@ export interface dxVectorMapOptions extends BaseWidgetOptions<dxVectorMap> {
        * @type object|Store|DataSource|DataSourceOptions|string|Array<any>
        * @notUsedInTheme
        */
-      dataSource?: any | Store | DataSource | DataSourceOptions | string;
+      dataSource?: CommonDataSource<any>;
```
Comment (dxBeardedBear) [inline js/viz/vector_map.d.ts]: Add object type

File: js/viz/vector_map.d.ts [hunk 10]
```diff
@@ -488,7 +486,7 @@ export interface dxVectorMapOptions extends BaseWidgetOptions<dxVectorMap> {
...
+    }> | { borderColor?: string; borderWidth?: number; color?: string; colorGroupingField?: string; colorGroups?: Array<number>; customize?: ((elements: Array<MapLayerElement>) => any); dataField?: string; dataSource?: CommonDataSource<any>; elementType?: 'bubble' | 'dot' | 'image' | 'pie'; hoverEnabled?: boolean; hoveredBorderColor?: string; hoveredBorderWidth?: number; hoveredColor?: string; label?: { dataField?: string; enabled?: boolean; font?: Font }; maxSize?: number; minSize?: number; name?: string; opacity?: number; palette?: Array<string> | PaletteType; paletteSize?: number; selectedBorderColor?: string; selectedBorderWidth?: number; selectedColor?: string; selectionMode?: 'multiple' | 'none' | 'single'; size?: number; sizeGroupingField?: string; sizeGroups?: Array<number>; type?: 'area' | 'line' | 'marker' };
```
Comment (dxBeardedBear) [inline js/viz/vector_map.d.ts]: Add object type

File: js/data/data_source_aliases.d.ts [hunk 11]
```diff
@@ -0,0 +1,8 @@
+import DataSource, { DataSourceOptions } from './data_source';
+import Store from './abstract_store';
+
+/**
+ * @docid
+ * @type Store|DataSource|DataSourceOptions|string|Array<any>
+ * */
+export type CommonDataSource<T> = string | Array<T> | Store<T> | DataSourceOptions<T> | DataSource<T>;
```
Comment (IlyaKhD) [inline js/data/data_source_aliases.d.ts]:
```suggestion
export type DataSourceLike<T> = string | Array<T> | Store<T> | DataSourceOptions<T> | DataSource<T>;
```

File: js/data/data_source_aliases.d.ts [hunk 12]
```diff
@@ -0,0 +1,8 @@
+import DataSource, { DataSourceOptions } from './data_source';
```
Comment (IlyaKhD) [inline js/data/data_source_aliases.d.ts]: I suggest moving this to the [js/data/data_source.d.ts](https://github.com/DevExpress/DevExtreme/blob/21_2/js/data/data_source.d.ts) module

File: js/core/index.d.ts [hunk 13]
```diff
@@ -22,3 +22,6 @@ export type Xor<T1, T2 = never, T3 = never, T4 = never, T5 = never, T6 = never,
 export type DeepPartial<T> = {
   [P in keyof T]?: DeepPartial<T[P]>;
 };
+
+// Omit does not exist in TS < 3.5.1
+export type Except<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
```
Comment (IlyaKhD) [inline js/core/index.d.ts]:
```suggestion
export type Skip<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
```
`Skip` seems to be more consistent with `Pick`, `Exclude`, `Omit`

File: js/data/data_source_aliases.d.ts (hunk 11 above)
Comment (IlyaKhD) [reply js/data/data_source_aliases.d.ts]: `Like` suffix is used widely in the Typescript internals in similar cases.

File: js/data/data_source_aliases.d.ts (hunk 12 above)
Comment (IlyaKhD) [reply js/data/data_source_aliases.d.ts]: [js/data/index.d.ts](https://github.com/DevExpress/DevExtreme/blob/21_2/js/data/index.d.ts) is even a better place for it, I think

Comment (alexserov) [reply js/data/data_source_aliases.d.ts]: moved to `/data/data_source.d.ts` to avoid circular deps

**Review by IlyaKhD — APPROVED** [review]

**Pull request merged**